### PR TITLE
Multitenancy

### PR DESCRIPTION
**Pull request by eduvilar**
References
===================
Issue #2611 

Objectives
===================
Make Consul Multitenant

Visual Changes
===================
The name of the current tenant appears in the main administration panel:
multitenant_01

From Multitenancy we can manage the different tenants:
multitenant_02

The following image shows the data that can be managed from each tenant:
multitenant_03

Notes
===================
*  Only the main tenant can manage the rest of the tenants.
*  The main tenant can not be deleted, and the subdomain can not be modified.
*  Can not access the administration panel of other tenants through the main tenant. For it, you have to change the tenant by subdomain, for now.

Pending questions
===================
*  Tenants switcher to access the administration panel of each one of them through the main tenant.

### CODE REVIEW

**Comment by voodoorai2000** [comment]
@eduvilar this is fantastic! 🎉 
Thank you so much! 👏 

This is a very important feature, and like all others, we must thoroughly test it 👍 
Would you mind adding specs for all these great functionalities? 😌

File: app/views/admin/tenants/_errors.html.erb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (aitbw) [inline app/views/admin/tenants/_errors.html.erb]: Please, make sure to remove this trailing whitespace :smile_cat:

File: app/models/tenant.rb [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+          populate_settings
+        end
+      end
+    end
+
+    def destroy_tenant
+      Apartment::Tenant.drop(subdomain)
+    end
+
+    def populate_settings
```
Comment (aitbw) [inline app/models/tenant.rb]: :thinking: why do we need to populate the DB through the model? Can't we do that with a Rake task?

File: app/controllers/admin/tenants_controller.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  private
+
+    def main_tenant?(tenant)
+      tenant.subdomain == 'public'
+    end
+
+    def tenant_params
+      params.require(:tenant).permit(:name, :title, :subdomain, :postal_code, :endpoint_census, :institution_code_census, :portal_name_census, :user_code_census)
```
Comment (aitbw) [inline app/controllers/admin/tenants_controller.rb]: This line is too long acc. to `rubocop`, can you adjust it?

File: app/controllers/admin/tenants_controller.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    if @tenant.save
+      redirect_to admin_tenants_path
+    else
+      render :new
+    end
+  end
+
+  def update
+    if @tenant.update(tenant_params)
+      redirect_to admin_tenants_path
```
Comment (aitbw) [inline app/controllers/admin/tenants_controller.rb]: Please, include a `notice` response for this action

File: app/controllers/admin/tenants_controller.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def edit
+  end
+
+  def create
+    @tenant = Tenant.new(tenant_params)
+
+    if @tenant.save
+      redirect_to admin_tenants_path
```
Comment (aitbw) [inline app/controllers/admin/tenants_controller.rb]: Please, include a `notice` response for this action

**Comment by eduvilar** [comment]
Ok @voodoorai2000 , I'll add specs for these functionalities. Also I expect you tell me that things can be improved.

File: app/controllers/admin/tenants_controller.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
+class Admin::TenantsController < Admin::BaseController
+
+  respond_to :html
```
Comment (aitbw) [inline app/controllers/admin/tenants_controller.rb]: Considering all responses are HTML (at least for this controller), do we really need this line?

File: Gemfile [hunk 7]
```diff
@@ -54,6 +54,7 @@ gem 'unicorn', '~> 5.4.0'
 gem 'whenever', '~> 0.10.0', require: false
 gem 'globalize', '~> 5.0.0'
 gem 'globalize-accessors', '~> 0.2.1'
+gem 'apartment'
```
Comment (aitbw) [inline Gemfile]: Please, include pessimistic versioning for this gem: `gem 'apartment', '~> 2.2'`

File: app/controllers/admin/base_controller.rb [hunk 8]
```diff
@@ -4,11 +4,23 @@ class Admin::BaseController < ApplicationController
...
+  before_action :get_tenants
 
   private
 
     def verify_administrator
       raise CanCan::AccessDenied unless current_user.try(:administrator?)
     end
 
+    def set_current_tenant
+      if session[:current_tenant] == nil
```
Comment (aitbw) [inline app/controllers/admin/base_controller.rb]: Replace `== nil` with `.nil?` :smile_cat:

**Comment by aitbw** [comment]
Oh, forgot to mention: There's no need to mention the issue number on the PR's title —by including it into the description is more than enough, so you can edit the title accordingly :smile_cat:

File: app/models/tenant.rb (hunk 2 above)
Comment (eduvilar) [reply app/models/tenant.rb]: We need to load data configuration when creating a new tenant, but I agree with you that it's not the best way to do it. I'll try to do it as you suggested.

**Comment by aitbw** [comment]
Hey @eduvilar, I noticed your branch has a ton of clean `merge` commits —is it possible you can `git rebase origin/master` on your local branch (that is, if your local `master` branch is up to date) and then `git push --force`?

This way we can ensure a clean Git history :raised_hands: what do you think?

**Comment by sbernhard** [comment]
Would be great, if the PR can be updated and merged finally when done.

**Comment by javierm** [comment]
@sbernhard Currently the changes in this pull request work great for new CONSUL installations, but testing shows they might break existing installations, making it impossible to access their existing data.

So I'm afraid we can't merge it until that issue is solved :relieved:. However, if you're working on a new installation, it's probably safe to apply these changes to your repository.

By the way, contributions are welcome :smile:. Feel free to have a look!

**Comment by sbernhard** [comment]
Nice. I guess it would make sense to create a plan / create some issues what needs to be done to merge this PR. If a upgrade strategy needs to be created, maybe someone would work on it if it's clear what needs to be adjusted.

**Comment by javierm** [comment]
I've rebased this pull request against our current master branch and opened pull request #4030. @eduvilar Thanks a lot for your contibutions! :wink: We'll continue to work on multitenancy in pull request #4030.

**Pull request closed without merging**